fenicsadapter/fenicsadapter.py

```diff
@@ -1,756 +1,379 @@
+        Sets up the coupling mesh. This function is called by initalize() function at the
         beginning of the simulation.
-
-        :param read_function_init: function on the read field
-        """
-        self._read_function_type = determine_function_type(read_function_init)
-        self._read_data = self._convert_fenics_to_precice(read_function_init)
-
-    def _create_coupling_boundary_condition(self):
-        """Creates the coupling boundary conditions using an actual implementation of CustomExpression."""
-        x_vert, y_vert = self._extract_coupling_boundary_coordinates()
-
-        try:  # works with dolfin 1.6.0
-            self._coupling_bc_expression = self._my_expression(element=self._function_space.ufl_element()) # elemnt information must be provided, else DOLFIN assumes scalar function
-        except (TypeError, KeyError):  # works with dolfin 2017.2.0
-            self._coupling_bc_expression = self._my_expression(element=self._function_space.ufl_element(), degree=0)
-        self._coupling_bc_expression.set_boundary_data(self._read_data, x_vert, y_vert)
-
-    def create_coupling_dirichlet_boundary_condition(self, function_space):
-        """Creates the coupling Dirichlet boundary conditions using
-        create_coupling_boundary_condition() method.
-
-        :return: dolfin.DirichletBC()
+        :param mesh: FEniCS mesh
+        :param: subdomain: Part of FEniCS mesh which which will be computed by this participant
         """
-        self._function_space = function_space
-        self._create_coupling_boundary_condition()
-        return dolfin.DirichletBC(self._function_space, self._coupling_bc_expression, self._coupling_subdomain)
+        fenics_vertices, self._coupling_mesh_vertices, self._n_vertices \
+            = extract_coupling_boundary_vertices(mesh, subdomain, self._fenics_dimensions, self._dimensions)
+
+        if self._n_vertices > 0:
+            self._vertex_ids = self._interface.set_mesh_vertices(self._mesh_id, self._coupling_mesh_vertices)
+            """ Define a mapping between coupling vertices and their IDs in preCICE """
+            id_mapping = dict()
+            for i in range(self._n_vertices):
+                id_mapping[fenics_vertices[i].global_index()] = self._vertex_ids[i]
+
+            if MPI.size(MPI.comm_world) == 1:  # nearest-projection is only supported for non-parallel runs
+                edge_vertex_ids1, edge_vertex_ids2 = extract_coupling_boundary_edges(mesh, subdomain, id_mapping)
+
+                """ Set mesh edges in preCICE to allow nearest-projection mapping"""
+                for i in range(len(edge_vertex_ids1)):
+                    assert (edge_vertex_ids1[i] != edge_vertex_ids2[i])
+                    self._interface.set_mesh_edge(self._mesh_id, edge_vertex_ids1[i], edge_vertex_ids2[i])
+            else:
+                print("Parallel FEniCS runs only support nearest neighbor.")
```

Comment:
Nearest projection is not supported for parallel runs. I decided to not investigate this feature in depth. Therefore it might be possible to provide connectivity information also for parallel runs quite easily. I do not really know, since I decided to ignore this feature for the first prototype.